extension/src/lib/captionsEnactor.js

```diff
@@ -0,0 +1,564 @@
+            if (that.model.captionsEnabled !== captionsEnabled) {
+                that.model.captionsEnabled = captionsEnabled;
+                that.setCaptions();
+            }
+        };
+
+        that.player = gpii.uioPlus.player.createYTPlayer(videoElm, {
+            events: {
+                onApiChange: that.setCaptions
+            }
+        });
+
+        return that;
+    };
+
+    /**
+     * Adds the "enablejsapi=1" query parameter to the query string at the end of the src attribute.
+     * If "enablejsapi" already exists it will modify its value to 1. This is required for API access
+     * to the embedded YouTube video.
+     *
+     * @param {jQuery|Element} videoElm - a reference to the existing embedded YouTube video.
+     */
+    gpii.uioPlus.player.enableJSAPI = function (videoElm) {
+        var url = new URL(videoElm.getAttribute("src"));
+
+        url.searchParams.set("enablejsapi", 1);
+        videoElm.setAttribute("src", url.toString());
+    };
+
+    gpii.uioPlus.player.createYTPlayer = function (videoElm, options) {
+        var id = gpii.uioPlus.allocateSimpleId(videoElm);
+        gpii.uioPlus.player.enableJSAPI(videoElm);
+        return new window.YT.Player(id, options);
+    };
+
+    /**********************************************************************************************
+     * Captions component
+     **********************************************************************************************/
+
+    gpii.uioPlus.captions = function () {
+        var that = {};
+
+        that.players = [];
+        that.model = {
+            captionsEnabled: false
+        };
+
+        // bind a window event to update the model and players
+        window.addEventListener("message", function (event) {
+            gpii.uioPlus.captions.updateFromMessage(that.model, event, function () {
+                gpii.uioPlus.captions.updatePlayers(that);
+            });
+        });
+
+        // wait for the YouTube iframe API and create players
+        if (window.YT && window.YT.Player) {
+            that.players = gpii.uioPlus.captions.createPlayers(that.model.captionsEnabled);
+        } else {
+            // the YouTube iframe api will call onYouTubeIframeAPIReady after the api has loaded
+            window.onYouTubeIframeAPIReady = function () {
+                that.players = gpii.uioPlus.captions.createPlayers(that.model.captionsEnabled);
+            };
+        }
+
+        return that;
+    };
+
+    gpii.uioPlus.captions.createPlayers = function (captionsEnabled) {
+        var players = [];
+        var videos = document.querySelectorAll("iframe[src^=\"https://www.youtube.com/embed/\"]");
+
+        videos.forEach(function (video) {
+            players.push(gpii.uioPlus.player(video, captionsEnabled));
+        });
+
+        return players;
+    };
+
+    gpii.uioPlus.captions.updateFromMessage = function (model, event, callback) {
+        var settings = event.data["UIO+_Settings"];
```

Comment:
Similarly here it will be easier to detect the event type from its "type" field and then decide what to do with it, rather than checking every message for every type of payload (see above)